# Question 1

Supposons qu'on lit le document numéro '422908' dans ma matrice. Appliquez l'algorithme Page rank pour déterminer les autres lectures recommandées. En plus de la simple recommandation des références de '422908', appliquez au moins une variation de cette approche de base, comme celle exposée en classe qui consiste à étendre le sous-ensemble S (références) à S' (références des références). Expliquez la démarche que vous avez prise.

Notre article:

422908
Title:
Symbolic Model Checking for Real-time Systems

Description:
. We describe finite-state programs over real-numbered time in a guardedcommandlanguage with real-valued clocks or, equivalently, as finite automata withreal-valued clocks. Model checking answers the question which states of a real-timeprogram satisfy a branching-time specification (given in an extension of CTL with clockvariables). We develop an algorithm that computes this set of states symbolically as afixpoint of a functional on state predicates, without constructing the state space.For this purpose, we introduce a -calculus on computation trees over real-numberedtime. Unfortunately, many standard program properties, such as response for allnonzeno execution sequences (during which time diverges), cannot be characterizedby fixpoints: we show that the expressiveness of the timed -calculus is incomparableto the expressiveness of timed CTL. Fortunately, this result does not impair thesymbolic verification of "implementable" real-time programs---those whose safety...

Authors:

Joseph Sifakis, Miniparc-zirst Rue Lavoisier, Sergio Yovine, Thomas A. Henzinger, Xavier Nicollin,

In [110]:
# Lecture de la base de données
m <- read.table('citeseer.rtable')
# Transformation de m en matrice pour effectuer les calculs
m <- as.matrix(m)
# On normalise la matrice et on enelve les NaN
m <- m/colSums(m)
m[which(!is.finite(m))] <- 0

# Lecture de la fonction de Page Rank
# Si la division du vecteur de PR et le vecteur de nombre de lien sortant donne une valeur infini, on le met alors à 1 sur
# sur la dimension de la matrice comme dit par M. DesMarais
page <- function(mat,d,pr) {y <- pr/colSums(mat); y[which(!is.finite(y))] <- 1/dim(mat)[1]; (1-d)/dim(mat)[1] + (d * (mat %*% (y)))} 

In [111]:
# Lecture de la fonction pour trouver l'indice max
max.nindex <- function(m, n=5) {
    i <- order(m, decreasing=TRUE)
    return(i[1:n])
}

In [113]:
# Algorithme de Page Rank pour les recommandations les plus populaires
# On fait une nouvelle matrice avec S
m.temp <- m

d <- 0.85
pr <- rep(1,dim(m.temp)[1])
mm <- matrix(0, dim(m.temp)[1], 100)

for (i in 1:100) {pr <- page(m.temp,d,pr) ; mm[,i] <- pr} 
max.nindex(mm[,10],5)
# le numéro de l'article
rownames(m.temp)[max.nindex(mm[,10],5)]

[1] 865 443 518 158 640

[1] "500980" "259812" "297641" "14606"  "356801"

En premier, nous avons appliquer l'algorithme de Page Rank sur l'ensemble de la matrice m. 

50098
Title:
FunState - An Internal Design Representation for Codesign

Description:
In this paper, an internal design model called FunState (functions driven by state machines) is presented that enables the representation of different types of system components and scheduling mechanisms using a mixture of functional programming and state machines. It is shown how properties relevant for scheduling and verification of specification models such as Boolean dataflow, cyclostatic dataflow, synchronous dataflow, marked graphs, and communicating state machines as well as Petri nets can be represented in the FunState model of computation. Examples of methods suited for FunState are described, such as scheduling and verification. They are based on the representation of the model's state transitions in form of a periodic graph. The feasibility of the novel approach is shown with an ATM switch example.

Authors:

Dirk Ziegenbein, Jurgen Teich, Karsten Strehl, Lothar Thiele, Matthias Gries, Rolf Ernst, Student Member,

On voit que l'article n'est pas relier a notre article, il retourne seulement un article qui est tres cite et est cite par des articles tres cite aussi. En regardant sur google scholar, cet article est cite par 89articles ce qui est relativement beaucoup. Ce n'est pas un bonne recommandation comme on pouvait s'y attendre

In [80]:
# S' 

#Calcul de la matice de transition Sprime afin de determiner les chemins transitifs entre les noeuds par la puissance de la matrice
m2 <- m %*% m
m.t <- t(m)


In [81]:
# Déterminer les valeurs qui son associé au document recherché dans la matice S et S' les pages référencées par les pages de s
new.indice <- m['422908',]>0 | m2['422908',]>0 | m.t['X422908',]>0
# Construire une nouvelle matrice qui contient tout les documents qui sont un lien direct ou transitif vers le doucment recherché
m.temp <- m+m2+m.t
m3 <- m.temp[new.indice,new.indice]

In [82]:
# Appliquer PR sur la matrice issue de S et S' en itérant 10 fois 
d <- 0.85
pr <- rep(1,dim(m3)[1])
mm2 <- matrix(0, dim(m3)[1], 10)
for (i in 1:10) {pr <- page(m3,d,pr) ; mm2[,i] <- pr} 
max.nindex(mm2[,10],5)

rownames(m3)[max.nindex(mm2[,10],5)]

[1] 29 38 42 23 17

[1] "297641" "392298" "422908" "226359" "177173"

297641
Title:
Automatic Verification of Real-time Systems with Discrete Probability Distributions

Description:
We consider the timed automata model of [3], which allows the analysis of realtimesystems expressed in terms of quantitative timing constraints. Traditional approachesto real-time system description express the model purely in terms of nondeterminism;however, it is often desirable to express the likelihood of the systemmaking certain transitions. In this paper, we present a model for real-time systemsaugmented with discrete probability distributions. Furthermore, two approaches tomodel checking are introduced for this model. The first uses the algorithm of [6]to provide a verification technique for our model against temporal logic propertieswhich can refer both to timing properties and probabilities. The second, generallymore efficient, technique concerns the verification of probabilistic, real-time reachabilityproperties.1 IntroductionThe proliferation of digital technology embedded into real-life environmentshas led to increased interest in computer systems expressed i...

Authors:

And Jeremy Sproston, Marta Kwiatkowska, Roberto Segala,

Cet article semble etre une bonne recommandation, il parle de systeme temps reel comme notre article initial. Ce n'est pas le meme auteur principal ce qui est bon signe. Il est cite 336 ce qui lui donne une bonne credibilite. On pouvait s'attendre a une article populaire car c'est souvent le resultat du pagerank. 

# Question 2

Comparez les résultats obtenus avec une approche basée sur la similarité des articles dans un espace vectoriel, à l'instar du calcul de similarité de l'approche item-item. La mesure de la similarité et la façon de l'utiliser pour estimer la pertinence d'articles similaires est laissé à votre discrétion.

Comme approche de similarite nous avons choisi le cosinus.

In [147]:
m = read.table("citeseer.rtable")
m <- as.matrix(m)
m.sparse <- m
ST_id<-grep("422908", rownames(m))

In [148]:
min.nindex <- function(m, n=5) {
    i <- order(m)
    return(i[1:n])
}
#Fonctions de cosinus 
cosinus.vm <- function(v,m) { 
    n <- sqrt(colSums(m^2)); 
    (v %*% m)/(n * sqrt(sum(v^2)))
}

In [149]:
# on calcule le cosinus 
wcos.st <- cosinus.vm(m.sparse[,ST_id], m.sparse)

In [150]:
# Par approche item-item, le document X124 est recommandé avec la valeur de cosinus max
max.nindex(wcos.st,6)[-1]
colnames(wcos.st)[max.nindex(wcos.st,6)[-1]]

[1] 1081  858  755  889  989

[1] "X96767"  "X496938" "X425638" "X522428" "X7099"

96767
Title:
Model-Checking in Dense Real-time

Description:
. Model-checking is a method of verifying concurrent systems in whicha state-transition graph model of the system behavior is compared with a temporallogic formula. This paper extends model-checking for the branching-time logic CTLto the analysis of real-time systems, whose correctness depends on the magnitudesof the timing delays. For specifications, we extend the syntax of CTL to allowquantitative temporal operators such as 93!5 , meaning "possibly within 5 timeunits." The formulas of the resulting logic, Timed CTL (TCTL), are interpretedover continuous computation trees , trees in which paths are maps from the set ofnonnegative reals to system states. To model finite-state systems we introduce timedgraphs --- state-transition graphs annotated with timing constraints.As our main result, we develop an algorithm for model-checking, for determiningthe truth of a TCTL-formula with respect to a timed graph. We argue that choosinga dense domain instead of a discrete domain to m...

Authors:

Costas Courcoubetis, David Dill, Rajeev Alur,

On trouve ici une recommandation qui est tres pres de notre article de base. En effet, en comparant les titres:

Model-Checking in Dense Real-time

Symbolic Model Checking for Real-time Systems

On s'attend a ce que l'article soit tres pertinant par rapport a notre article de base. Nous avons donc ici une bonne recommandation selon nous etant donne le fait que les auteurs ne soit pas les memes que notre article de base. L'approche cosinus semble une bonne approche de similarite.

Une vraie approche item-item est utilie pour voir la comparaison avec l'approche de distance

In [38]:
voisins <- 21
min.nindex <- function(m, n=5) {
    i <- order(m)
    return(i[1:n])
}
cosinus.vm <- function(v,m) { 
    n <- sqrt(colSums(m^2)); 
    (v %*% m)/(n * sqrt(sum(v^2)))
}

# moyenne des votes par utilisateur
u.mean <-rowMeans(m.sparse)
# on centre la matrice pour enlever le biais des utilisateurs qui votent toujours haut ou en bas
m.center <- as.matrix((m.sparse-u.mean))

# on remplace les 0 par NA
m.center[m.center == 0] <- NA
# on calcule la distance eucledienne entre les films
distance.na.450.center <- sqrt(colSums((m.center[,ST_id] - m.center)^2, na.rm=T)) # ignore les valeurs manquantes


# on calcule le nombre de votes communs des films avec star trek V
votes.communs <- (colSums((m.sparse[,ST_id] * m.sparse) > 0))

# on assigne la valeur Inf au film qui ont moins de 10 votes communs avec star trek V
# on filtre donc pour avoir 10 ou plus votes communs
distance.na.450.center[votes.communs<1] <- Inf

# les films les plus proches sont choisis (les 20 plus proches)
# on enlève Star Trek V encore une fois
closest.euc <- min.nindex(distance.na.450.center, voisins)[-1]
# on calcule le cosinus entre les films choisis et star trek V
wcos.st <- cosinus.vm(m.sparse[,ST_id], m.sparse[,closest.euc])

# les 20 films les plus communs
idx<- closest.euc
# on transpose m
m.sparse.t <- m.sparse
# la range de l item de star trek V
v.item <- m.sparse[ST_id,]
# la moyenne de l item de star trek V
v_1 <-mean(v.item);
# retourne la moyenne pour chacun des 20 films voisisns
v_i <- apply(m.sparse[idx,], 1, FUN=mean)

x<- as.matrix((m.sparse[idx,]))

# retourne un liste ayant comme element chaque colonne de x
new_x <-lapply(seq_len(ncol(x)), function(i) x[,i])
# retourne un liste ayant comme element la moyenne des 20 films voisins
new_y <-as.vector(lapply(seq_len(ncol(x)), function(i) v_i)) 
# retourne un liste ayant comme element les poids(cosinus) des 20 films voisins
new_w <-as.vector(lapply(seq_len(ncol(x)), function(i) as.matrix(wcos.st)))
#prend seulement ceux qui n'ont de cote 
# Fonction qui détermine la prédiction des valeurs manquants à l'aide de l'algorithme user-user ou item-item                        
predict_func <- function(x,y,w) {
    x.temp <- (x-y);
    num <-  w %*%x.temp;
    denum <- sum(abs(t(w)));
    num/denum} 

c_prod <- mapply(predict_func ,new_x, new_y, new_w)

                         
# Ajoute la moeynne de l'item pour la note finale des items
prediction <- v_1 + c_prod
# on retourne les resultats qui n'avaient pas de prediction
results <- prediction
rownames(m)[max.nindex(results,6) ][-1]

[1] "64835"  "70445"  "177173" "17094"  "426325"

On remarque que les recommandation ne sont pas les memes que pour l'approche distance avec cosinus.
On remarque aussi que sur les 5 des recommandations, 4 d'entre elles est du meme autour que notre article original. On s'attend a obtenir des recommandations avec les memes auteurs avec l'approche item item car les auteurs citent les memes references et se citent eux-meme.

# Question 3

Utilisez une validation croisée pour évaluer la performance de l'approche de similarite. Vous pouvez vous inspirer de l'approche utilisée dans l'article suivant pour la méthodologie à adopter. 

Pour la validation croisee on validera sur la methode cosinus

In [2]:
# Lecture de la fonction pour déterminer la matrice de Cosinus de m 
matrix.cos <- function(m) {
    (m %*% t(m)) / ( t(matrix(sqrt(rowSums(m^2)),nrow(m),nrow(m)))
                   * (sqrt(rowSums(m^2))) )
}

In [25]:
m <- read.table('citeseer.rtable')
m <- as.matrix(m)
# on met la diagonale proche de 0 pour evite les NaN
diag(m)<- 0.0001
# Trouver les indices des ariticles référencés 
indice <- which(m == max(m), arr.ind = TRUE);
# Resample la matrice m 
set.seed(3)
len <- 100
indice<-indice[sample(len),]
vrai <- rep(len,0)
test <- rep(len,0)
mat.cos <- matrix.cos(m)

# pour des raisons de temps on mets a 100 le nombre de repetitions
for (i in 1:dim(indice)[1]) {
# On détermine les documents référencé dans la matrice
temp <- m ;
    
# On met à 0 le docuement pour effectuer la validation croisée
temp[indice[i,1],indice[i,2]] <- 0;
# On calcul ensuite la valeur de cosinus pour les n documents référencés
vrai[i]<-mat.cos[indice[i,1],indice[i,2]]
test[i]<-matrix.cos(temp)[indice[i,1],indice[i,2]]
    }


# Par la méthode décrite avec 10 itérations, on met à 0 un article référencé on analyse les impacts sur les 
# recommandations de ce dernier. 

In [114]:
mean(abs(vrai[vrai>0.05]-test[vrai>0.05]))

[1] 0.1598938

In [115]:
# rappel
sum(test>0.05)/sum(vrai>0.05)



[1] 0.765625

Nous avons opter pour une validation croisee leave one etant donnee le peu de recommandation que nous avions.
On enleve la recommandation par la suite on recalcule la valeur du cosinus pour cette article. On regarde uniquement les recommandations car le fait qu'un article ne soit pas recommander ne nous interesse pas dans le cadre de cette analyse(on regarde donc uniquement le rappel). 

On ne regarde pas la precison parce que ca ne ferait pas de sens de transformer une non reference en reference. On ne peut pas non plus le mettre a -1 car cela aura une influence sur le numerateur du cosinus qui pourra devenir negatif ce qui serait difficile a interpreter. 

On considere une recommandation quand son cosinus est plus grand que 0.05(arbitraire) pour qu'il ait un minimum de similarite  et aussi pour filtrer les effets de la diagonale, qui a ete mise a 0.0001, lors du calcul du cosinus. (de facon a eviter les NaN)

On regarde par la suite le rappel des predictions.

On voit qu'environ 25% des predictions sont perdus dans le processus de CV. Cela est a cause des situations ou il y a seulement un vote en commun entre les deux articles sont cosinus est alors mis a 0. 

On peut aussi calculer le MAE. Il est de 0.16 ce qui est assez elever etant donnee les cosinus qui vont jusqu'a environ 0.5.






Une meilleur approche selon nous serait plutot de regarder le variation sur le rang des recommandations.

Comme lu dans le papier recommander pour le TP2 nous allons proceder a une validation croisee de la sorte.
Le but de cet excercise est d'analyser les effets d'enlever une reference et de comparer les resultats sur le rang des autres articles referencés. On aurait pu repeter les memes procedure pour les autres articles, mais cela aurait demandé trop de temps de calcul, nous nous sommes donc concentrer sur notre article de départ et ses références pour regarder les varations de rang.

Le rang correspond a la position de la recommandation. (position 1 = meilleure recommandation)

In [116]:
cosinus.vm <- function(v,m) { 
    n <- sqrt(colSums(m^2)); 
    (v %*% m)/(n * sqrt(sum(v^2)))
}

In [212]:
# Par la méthode iétatif, on calcule les recommandations pour l'article 422908 en enlevant une référence à chaque fois 
ST_id <- 747
m <- read.table('citeseer.rtable')
m <- as.matrix(m)
m.sparse<-m
# calcul du cosinus original
wcos.st <- cosinus.vm(m[,ST_id], m)
# on utilise les indices du des refenreces de l'article
indice <- which(m.sparse[,ST_id]==1, arr.ind = TRUE)
len <- length(indice)

rank.T <- matrix(, nrow = len, ncol = len)

rank.N <- matrix(, nrow = len, ncol = len)

# on repete pour chaque reference
for (i in 1:len) {

m.sparse<-m
# on met une des references a 0
m.sparse[indice[i],ST_id]<-0
# on calcule le nouveau cosinus
wcos.st2 <- cosinus.vm(m.sparse[,ST_id], m.sparse)
    # on calcul le nouveau rang pour les autres refenreces
    for (j in 1:len){
        if (j!=i){
# on calcule le rang de la vraie refenrece pour toutes les autres refenrences
rank.T[i,j] <- which(colnames(wcos.st)[max.nindex(wcos.st,1090)]==colnames(m)[indice[j]],arr.ind=TRUE)
# on calcule le nouveau rang de la refenrece pour toutes les autres refenrences
rank.N[i,j] <- which(colnames(wcos.st2)[max.nindex(wcos.st2,1090)]==colnames(m)[indice[j]],arr.ind=TRUE)  
            }
        }
    }


In [205]:
article.rank <- rank.T[1,]
article.rank[1]<- rank.T[2,1]
# dans cette matrice on voit le rang inital de toutes les references de l'article de depart
article.rank

[1]   18   80   33   11  750   17   15   34   21   29   72   20   47   88  103
[16]  309    9   49  886   50  401    7   27   42  441  962  498  500   55   63
[31]   92    3   12   22   93    8   78    6 1058   79   58  678 1065   23    2
[46]  708

In [206]:
# ici on voit le changement de rang pour chacune des situations
abs(rank.N-rank.T)

[,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
 [1,] NA    4    0    0    0    1    1    0    1    1    ...     3    1   
 [2,]  3   NA    0    4    0    2    2    0    4    2    ...     4    5   
 [3,]  1    0   NA    3    0    1    2    0    0    0    ...     0    5   
 [4,]  0    0    0   NA    0    0    2    0    5    0    ...     0    1   
 [5,]  1    0    0    0   NA    1    0    0    1    1    ...     0    1   
 [6,]  1    0    0    7    0   NA    3    0    0    0    ...     0    2   
 [7,]  0    0    0    1    0    1   NA    0    5    0    ...     0    0   
 [8,]  1    2    0    1    0    1    1   NA    0    0    ...     2    0   
 [9,]  0    0    0    0    0    1    1    0   NA    0    ...     0    1   
[10,]  3    2    1    1    0    4    1    1    2   NA    ...     2    0   
[11,]  0    1    1    1    0    1    1    1    0    0    ...     1    7   
[12,]  3    1    0    5    0    4    4    0    5    0    ...     1    2   
[13,]  3    6   45    3    0    3    8   45    3   25    ...     6    5   
[14,]  4    0    1    0    0    2    9    1    3    0    ...     0    0   
[15,]  1    0    0    0    0    1    0    0    1    1    ...     0    1   
[16,]  1    2    1    1    0    1    1    1    0   29    ...     2    0   
[17,]  5    0    1    1    0    1    1    1    1    2    ...     0    6   
[18,]  1    2    0    0    0    1    1    0    0    0    ...     2    0   
[19,]  2    2    1    0    0    5    1    1    1    1    ...     2    0   
[20,]  0    8    0    1    0    0    1    0    0    0    ...     8    0   
[21,]  0    2    0    0    0    0    0    0    0    0    ...     2    0   
[22,]  2    1    0    6    0    2    2    0    0    0    ...     1    1   
[23,]  2    2    0    6    0    3    4    0    0    0    ...     2    0   
[24,]  2    1    0    6    0    5    3    0    1    1    ...     1    1   
[25,]  1    1    1    0    0    1    8    1    1    2    ...     1    0   
[26,]  0    6    0    0    0    0    0    0    0    0    ...     6    0   
[27,]  2   24    0    6    0    4    3    0    2    0    ...     4    0   
[28,]  0    1    2    0    0    0    0    2    1    1    ...     1    1   
[29,]  2    3    0    0    0    2    9    0    4    0    ...     3    1   
[30,]  1    4   47    2    0    1    2    2    0    1    ...   566    5   
[31,]  0    0    0    0    0    1    1    0    0    0    ...     0    0   
[32,]  1    0    0    1    0    1    2    0    0    0    ...     0    1   
[33,]  0    0    0    0    0    1    1    0    0    1    ...     0    1   
[34,]  0    1    0    0    0    1    1    0    0    1    ...     1    0   
[35,]  5    1    1    1    0    1    1    1    2   31    ...     1    6   
[36,]  0    0    0    0    0    0    0    0    1    0    ...     0    0   
[37,]  0    1    1    2    0    0    1    1    0    1    ...    NA    6   
[38,]  0    0    1    0    0    0    0    1    1    0    ...     0   NA   
[39,]  2   24    0    1    0    4    2   47    2    0    ...     2    0   
[40,]  1    3    0    0    0    1    1    0    1    1    ...     3    1   
[41,]  0    0    1    0    0    0    0    1    0    1    ...     0    1   
[42,]  1    1    1    1    0    5    0    1    1    0    ...     1    7   
[43,]  6    1    2    0    0    0    0    2    1    1    ...     1    0   
[44,]  1    0    0    1    0    2    3    0    5    0    ...     0    0   
[45,]  0    0    0    0    0    1    1    0    0    1    ...     0    1   
[46,]  0    0    0    0    0    0    0    0    5    0    ...     0    0   
      [,14] [,15] [,16] [,17] [,18] [,19] [,20] [,21]
 [1,]  0    586     1    0     0     1     0     0   
 [2,]  0      4     2    0     0     6     0     0   
 [3,]  0      0     0    0     0     0     0     0   
 [4,]  0      0     0    0     0     1     0     0   
 [5,]  0      0     0    0     0     1     0     0   
 [6,]  0      0     0    0     0     1     0     0   
 [7,]  0      0     0    0     0     1     0     0   
 [8,]  0      2     1    0     0     0     0     

In [207]:
# la moyenne de changment de rang pour chacune des references
rowMeans(abs(rank.N-rank.T),na.rm=T)

[1] 13.9555556  3.2444444  0.7555556  0.5111111  3.6000000  0.7777778
 [7]  0.3111111  1.1555556  0.4666667  1.2444444  0.8888889  1.8444444
[13]  5.3333333  0.8444444  0.3111111  2.4888889  0.8888889  0.6444444
[19]  1.2000000  2.2000000  0.9777778  0.6000000  0.8444444  1.0444444
[25]  1.8222222  9.0666667  7.2444444  1.9333333 15.0666667 16.0666667
[31]  0.1777778  0.5333333  0.1555556  0.1777778  1.7111111  0.1777778
[37]  4.2000000  0.3777778  2.8000000  0.7333333 10.7111111  1.6444444
[43] 13.2444444  0.4888889  0.4222222  0.1555556

In [208]:
# la moyenne globale de changement de rang
mean(rowMeans(abs(rank.N-rank.T),na.rm=T))

[1] 2.935749

On remarque qu'avec le cosinus meme si une refence est enleve on retrouvera cette recommandation dans le meme rang environ que si elle n'etait pas la. En effet, la moyenne du changement de rang est de 3. L'approche du cosinus recommandera cette article dans la majorite des cas a la meme position ou dans une position tres proche meme si la reference de l'article est enleve. 



Cette moyenne pourrait etre diminuer en se limitant au articles recommandable a la base.
si on garde les colonnes avec des articles qui ont un rang inital <= 50 on devrait trouver un meilleur resultat.

In [213]:
abs(rank.N[,article.rank<=50]-rank.T[,article.rank<=50])

[,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
 [1,] NA    0    0    1    1    0    1    1    1    1    ...    1     2   
 [2,]  3    0    4    2    2    0    4    2    3    1    ...    2     2   
 [3,]  1   NA    3    1    2    0    0    0    0    0    ...    1     1   
 [4,]  0    0   NA    0    2    0    5    0    0    0    ...    9     0   
 [5,]  1    0    0    1    0    0    1    1    1    0    ...    1     2   
 [6,]  1    0    7   NA    3    0    0    0    0    0    ...    8     1   
 [7,]  0    0    1    1   NA    0    5    0    0    0    ...    0     0   
 [8,]  1    0    1    1    1   NA    0    0    0    1    ...    0     1   
 [9,]  0    0    0    1    1    0   NA    0    0    1    ...    1     2   
[10,]  3    1    1    4    1    1    2   NA    2    2    ...    0     1   
[11,]  0    1    1    1    1    1    0    0    0    2    ...    1     1   
[12,]  3    0    5    4    4    0    5    0   NA    1    ...    7     0   
[13,]  3   45    3    3    8   45    3   25    3   NA    ...    2     2   
[14,]  4    1    0    2    9    1    3    0    5    1    ...    0     0   
[15,]  1    0    0    1    0    0    1    1    1    0    ...    1     2   
[16,]  1    1    1    1    1    1    0   29    0   40    ...    0     0   
[17,]  5    1    1    1    1    1    1    2    1    1    ...    1     2   
[18,]  1    0    0    1    1    0    0    0    0    0    ...    0     0   
[19,]  2    1    0    5    1    1    1    1    1    2    ...    0     1   
[20,]  0    0    1    0    1    0    0    0    0    1    ...    0     0   
[21,]  0    0    0    0    0    0    0    0    0    1    ...    0     0   
[22,]  2    0    6    2    2    0    0    0    0    0    ...   NA     0   
[23,]  2    0    6    3    4    0    0    0    0    1    ...    0    NA   
[24,]  2    0    6    5    3    0    1    1    1    0    ...    8     1   
[25,]  1    1    0    1    8    1    1    2    1    1    ...    0    45   
[26,]  0    0    0    0    0    0    0    0    0    1    ...    0     0   
[27,]  2    0    6    4    3    0    2    0    5    3    ...    8     1   
[28,]  0    2    0    0    0    2    1    1    5    2    ...    1    46   
[29,]  2    0    0    2    9    0    4    0    1    0    ...    1     1   
[30,]  1   47    2    1    2    2    0    1    0   39    ...    1     1   
[31,]  0    0    0    1    1    0    0    0    0    0    ...    0     1   
[32,]  1    0    1    1    2    0    0    0    0    0    ...    8     0   
[33,]  0    0    0    1    1    0    0    1    0    0    ...    1     2   
[34,]  0    0    0    1    1    0    0    1    0    0    ...    0     1   
[35,]  5    1    1    1    1    1    2   31    2    1    ...    1     2   
[36,]  0    0    0    0    0    0    1    0    5    0    ...    0     0   
[37,]  0    1    2    0    1    1    0    1    0    0    ...    1     1   
[38,]  0    1    0    0    0    1    1    0    5    1    ...    0     0   
[39,]  2    0    1    4    2   47    2    0    5    1    ...    0     0   
[40,]  1    0    0    1    1    0    1    1    1    1    ...    1     2   
[41,]  0    1    0    0    0    1    0    1    0    1    ...    1     1   
[42,]  1    1    1    5    0    1    1    0    1    1    ...    1     0   
[43,]  6    2    0    0    0    2    1    1    1    3    ...    0     0   
[44,]  1    0    1    2    3    0    5    0    0    1    ...    0     0   
[45,]  0    0    0    1    1    0    0    1    0    0    ...    1     3   
[46,]  0    0    0    0    0    0    5    0    0    0    ...    0     0   
      [,14] [,15] [,16] [,17] [,18] [,19] [,20] [,21]
 [1,]  1     0     0     1     1     1     1     0   
 [2,] 28     0     2     4     2     5     6     0   
 [3,]  0     0     3     0    11     5     0     0   
 [4,]  0     0     2     1     1     1     1     0   
 [5,]  0     0     0     1     0     1     1     0   
 [6,]  0     0     3     6     2     2     1     0   
 [7,]  0     0     1     1     0     0     1     0   
 [8,]  0     0     1     0    11     0     0     

In [214]:
rowMeans(abs(rank.N[,article.rank<=50]-rank.T[,article.rank<=50]),na.rm=T)

[1]  0.7272727  3.2608696  1.5454545  1.0454545  0.4782609  1.5909091
 [7]  0.4545455  0.9545455  0.7727273  1.7272727  1.3478261  3.3636364
[13]  8.4545455  1.5217391  0.4782609  4.1739130  1.6363636  0.4090909
[19]  1.2173913  0.6363636  0.1304348  0.8636364  1.2272727  1.9090909
[25]  3.2173913 14.5652174  3.4347826  3.0000000  1.3478261  5.3913043
[31]  0.1304348  1.0000000  0.3181818  0.1818182  2.8260870  0.3636364
[37]  1.0869565  0.6363636  3.3478261  0.7391304  0.7391304  1.1739130
[43]  1.1739130  0.8181818  0.6363636  0.3043478

In [215]:
mean(rowMeans(abs(rank.N[,article.rank<=50]-rank.T[,article.rank<=50]),na.rm=T))

[1] 1.877384

On voit que la moyenne de changment du rang est presque diminuer de moitie en gardant uniquement les recommandations qui pourrait etre considere pertinante.

Donc, pour les recommandatinos pertinantes, le rang change tres peu en moyenne ce qui est bon signe car la methode des cosinus nous permet encore de trouver de bonnes recommandations.